log/record_test.go

```diff
@@ -149,16 +149,40 @@ func TestRecordAllocationLimits(t *testing.T) {
 		body = r.Body()
 	}), "Body")
 
-	attrVal := []log.KeyValue{log.Bool("k", true), log.Int("i", 1)}
-	assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
-		var r log.Record
-		r.AddAttributes(attrVal...)
-		n = r.AttributesLen()
-		r.WalkAttributes(func(kv log.KeyValue) bool {
-			attr = kv
-			return true
-		})
-	}), "Attributes")
+	t.Run("Attributes", func(t *testing.T) {
+		attrVal := []log.KeyValue{
+			log.Bool("b1", true),
+			log.Int("i1", 324),
+			log.Float64("f1", -230.213),
+			log.String("s1", "value1"),
+			log.Map("m1", log.Slice("slice1", log.BoolValue(true))),
+			log.Bool("b2", false),
+			log.Int("i2", 39847),
+			log.Float64("f2", 0.382964329),
+			log.String("s2", "value2"),
+			log.Map("m2", log.Slice("slice2", log.BoolValue(false))),
+		}
+
+		assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
+			var r log.Record
+			r.AddAttributes(attrVal[:5]...)
+			n = r.AttributesLen()
+			r.WalkAttributes(func(kv log.KeyValue) bool {
+				attr = kv
+				return true
+			})
+		}), "5 attributes")
+
+		assert.LessOrEqual(t, testing.AllocsPerRun(runs, func() {
+			var r log.Record
+			r.AddAttributes(attrVal...)
+			n = r.AttributesLen()
+			r.WalkAttributes(func(kv log.KeyValue) bool {
+				attr = kv
+				return true
+			})
+		}), 1.0, "10 attributes")
```

Comment:
I see that there are 2 allocations.

```
    --- FAIL: TestRecordAllocationLimits/Attributes (0.00s)
        record_test.go:176: 
            	Error Trace:	/home/runner/work/opentelemetry-go/opentelemetry-go/log/record_test.go:176
            	Error:      	"2" is not less than or equal to "1"
            	Test:       	TestRecordAllocationLimits/Attributes
            	Messages:   	10 attributes
```

Do you know where the second comes from?

Reply:
Hmm, I don't. What version of Go are you using to test this? What system are you on?

Reply:
Err, you mean [this](https://github.com/open-telemetry/opentelemetry-go/actions/runs/8008404729/job/21874772548?pr=4957), right?

Reply:
I can just bump this to `2` or try to finagle with the `flag` package to determine if the `race` flag is passed. What do you think?

Reply:
The `flag` parsing option sounds buggy. If there is another flag that also adds an allocation (guessing `-gcflags` will likely have similar issues) we would have to add those as well.

Reply:
I'm also guessing increasing this to `2` could also be buggy given we are really testing the compilers ability to allocate slices efficiently here. If there is another compiler used it could require even more than `2` allocations.

Reply:
I think it will be better to have only zero allocation test cases. Others can  be (more) flaky.

Reply:
Btw you can consider creating create a assertion function named `assertZeroAlloc`

Reply:
Reverted. PTAL

Reply:
> Btw you can consider creating create a assertion function named `assertZeroAlloc`

I'm not sure we will be able to annotate as richly if we did that. Nor, would be gain much in conciseness given `assertZeroAlloc` will still need to take the anonymous func definitions.

I can take another look though if you still think it is worth while.

Reply:
```go
func assertZeroAllocs(t *testing.T, f func(), msgAndArgs ...interface{}) {
	t.Helper()
	assert.Equal(t, 0.0, testing.AllocsPerRun(5, f, msgAndArgs...)
}
```
I suggest doing it as a separate PR so that more people also review it and merge this PR as it is.